### PR TITLE
SDK - Compiler - Added support for volume-based data passing

### PR DESCRIPTION
**Pull request by Ark-kun**
Currently artifact passing is performed by Argo sidecar containers what download input data and upload output data to artifact repository (usually, S3-compatible blob storage like Minio).
The performance of this method is not optimal and it requires that pod disks have enough capacity to hold all artifact data.

This commit adds support for volume-based data passing.
This method involves using a single milti-write Kubernetes data volume to pass all intermediate data.
Parts of the volume are mounted to the input/output artifact directories, so when the user program reads and writes files, the files actually reside in the data volume.
This method improves the performance and reduces storage resource requirements.

The data volume must exist and support "READ_WRITE_MANY". (Or it can be an auto-provisioning PVC)

Limitations:
* All artifact file names must be the same (e.g. "data"). All auto-generated paths are already consistent. Avoid using any hard-coded paths.
* Passing constant values (text) as arguments for artifact inputs is not supported. (Rare case)
* The feature is experimental.

### CODE REVIEW

**Comment by Bobgy** [comment]
Some general questions:
* I remember GKE doesn't have MULTI_READ_WRITE offering, is that true?
* When we don't have MULTI_READ_WRITE, will the pipeline just fail, or will it run sequentially instead, (because only one step can mount that volume, other parallel steps will keep pending for that volume until it finishes)

**Comment by Ark-kun** [comment]
> * I remember GKE doesn't have MULTI_READ_WRITE offering, is that true?

Any NFS service supports MULTI_READ_WRITE. One managed example is Google Cloud FileStore. There are other options as well. For example, in Kubernetes it's pretty easy to expose any MULTI_READ_ONCE volume or mounted GCS volume as a MULTI_READ_WRITE NFS service.

>When we don't have MULTI_READ_WRITE, will the pipeline just fail, or will it run sequentially instead, (because only one step can mount that volume, other parallel steps will keep pending for that volume until it finishes)

It's in between: A sequential pipeline will work, but a parallel pipeline will fail.

>Why would that have better performance than artifact downloading.

It can be slightly faster since the component code can start immediately instead of waiting for the data to be downloaded. Some programs can do I/O prefetch reads in parallel with the CPU-intensive threads. It can also be faster if not all the data is used.

**Comment by Bobgy** [comment]
Thanks! That clears my confusion!

**Comment by Bobgy** [comment]
> It's in between: A sequential pipeline will work, but a parallel pipeline will fail.

I forgot to ask about this, I observed Kubernetes default behavior when a volume cannot be mounted seems to be pending for the volume to be available without a timeout. Therefore, sounds like a parallel pipeline will be turned into sequential automatically. Did you try this? Or maybe my assumption about Kubernetes isn't true here?

**Comment by Ark-kun** [comment]
> I observed Kubernetes default behavior when a volume cannot be mounted seems to be pending for the volume to be available without a timeout. Therefore, sounds like a parallel pipeline will be turned into sequential automatically. Did you try this? Or maybe my assumption about Kubernetes isn't true here?

Interesting idea. I should check this.

File: sdk/python/kfp/compiler/_data_passing_using_volume.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
+#/bin/env python3
+
```
Comment (rmgogogo) [inline sdk/python/kfp/compiler/_data_passing_using_volume.py:2]: Thanks Alexy. 

I thought user already can survive with existing SDK (option-1, pipelinevolume, option-2 volume/volumeMount). User's codes can pass string typed volume based file path cross components.

I will setup a meeting with you to quickly sync on this one. I may didn't well catch the reason on why need this PR due to my less knowledge on SDK.

**Comment by rmgogogo** [comment]
hi Alexy, not sure whether you already have a sample for this change to let me and other catch it.

https://github.com/kubeflow/pipelines/issues/3385

File: sdk/python/tests/compiler/testdata/artifact_passing_using_volume.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+
+volume_based_data_passing_method = data_passing_methods.KubernetesVolume(
+    volume=V1Volume(
+        name='data',
+        persistent_volume_claim=V1PersistentVolumeClaim(
+            spec=V1PersistentVolumeClaimVolumeSource(
+                claim_name='data-volume',
+            ),
+        ),
```
Comment (elikatsis) [inline sdk/python/tests/compiler/testdata/artifact_passing_using_volume.py:33]: You are very close :smile: `V1PersistentVolumeClaimVolumeSource` should be used directly
```suggestion
        persistent_volume_claim=V1PersistentVolumeClaimVolumeSource(
                claim_name='data-volume',
        ),
```

Comment (Ark-kun) [reply sdk/python/tests/compiler/testdata/artifact_passing_using_volume.py]: Yeah. I've fixed the YAML, but made a mistake here.

**Comment by Ark-kun** [comment]
> hi Alexey, not sure whether you already have a sample for this change to let me and other catch it.

A sample (compiler test) is included in this change: https://github.com/kubeflow/pipelines/pull/3371/files#diff-3768d2994e067512c65961f94257a3b2

The idea of this feature is: You take a pipeline and without any changes to the pipeline code configure it to use arbitrary Kubenetes volume for data passing (you can even use GCSFuse exposed as NFS volume).

When volume is used, the data is simply mounted instead of being uploaded and downloaded.

**Comment by Ark-kun** [comment]
We need to upgrade the Argo CLI in Travis first...

**Comment by Ark-kun** [comment]
Here is an example run of a pipeline using this feature.
image
There are no Argo artifacts (apart from the logs) - everything is stored in the volume.

**Comment by Ark-kun** [comment]
Ideally, the rewriter should be on the backend side (not in SDK) and the storage volume should be also used by the Frontend for artifact visualizations.

**Comment by rmgogogo** [comment]
/lgtm

Hi Alexey, here LGTM first to unblock your work/plan. I will send an email to you for volume-based data passing in TFX SDK etc.

**Comment by rmgogogo** [comment]
> > hi Alexey, not sure whether you already have a sample for this change to let me and other catch it.
> 
> A sample (compiler test) is included in this change: https://github.com/kubeflow/pipelines/pull/3371/files#diff-3768d2994e067512c65961f94257a3b2
> 
> The idea of this feature is: You take a pipeline and without any changes to the pipeline code configure it to use arbitrary Kubenetes volume for data passing (you can even use GCSFuse exposed as NFS volume).
> 
> When volume is used, the data is simply mounted instead of being uploaded and downloaded.
> 
> One interesting side effect feature that this feature enables is that the data is visible as it's being written. So it's possible to have TensorBoard that watches the training while it's being performed. Or you could do data streaming. (Both features are just ideas for future).

For TensorBoard case, I requested the feature to support dynamic logdir. If it can be implemented, we can just launch one TB instance. Besides this PR, we also can early launch TB instance during training as MLMD may provide artifact type for it. +1 to your comment it also can helps.

**Pull request merged**